content/blogs/gsoc-memory-disclosure-week-1.md

```diff
@@ -0,0 +1,69 @@
+---
+title:       "Detect kernel information disclosure by Bochspwn-reloaded - GSoC 2020 - First week"
+author:      "khanhnt"
+date:        2020-06-03
+tags:        [ gsoc ]
+---
+
+## Introduction
+
+Hello, I am Nguyen Trung Khanh (@khanhnt) from Vietnam and I am one of the GSoC students of ReactOS.
+
+My project is running ReactOS on [bochspwn-reloaded](https://github.com/googleprojectzero/bochspwn-reloaded) to list and fix all the bugs which were found by the tool. Additional, I have a week to implement detection of uninitialized memory use.
+
+## The first week
+
+Before GSoC, I did compile bochspwn-reloaded and run ReactOS on it so my work in the first week is pretty easy. I added some lines of code into bochspwn-reloaded to change its stack trace format, now it looks like windbg stack trace. This feature was added from commit [e7a912897](https://github.com/reactos/bochspwn-reloaded/commit/e7a912897aeffaa70e0bf1fd2bf590a4f65594dc).
+
+There are 2 kinds of functions when they were compiled: inline and non-inline. Inline function is the function that has `inline` declaration keyword or template function and its assembly code will be inlined at the call of caller function. For non-inline function, caller will use `call` assembly instruction to jump to address of non-inline function.
+
+Therefore, I use 2 functions [SymGetLineFromInlineContext](https://docs.microsoft.com/en-us/windows/win32/api/dbghelp/nf-dbghelp-symgetlinefrominlinecontext) and [SymGetLineFromAddr](https://docs.microsoft.com/en-us/windows/win32/api/dbghelp/nf-dbghelp-symgetlinefromaddr) to get source code information from given address.
+
+This is the result:
+
+```
+Stack trace:
+ #0  0x805817bd ((001817bd) ntoskrnl.exe!memcpy+3d)
+ #1  0x8054e231 ((0014e231) ntoskrnl.exe!KeUserModeCallback+f1 [h:\project\reactos\ntoskrnl\ke\i386\usercall.c @ 162])
+ #2  0xf7603dda ((00036dda) win32k.sys!co_IntGetCharsetInfo+da [h:\project\reactos\win32ss\user\ntuser\callback.c @ 1075])
+ #3  0xf7639bd6 ((0006cbd6) win32k.sys!UserLoadKbdLayout+236 [h:\project\reactos\win32ss\user\ntuser\kbdlayout.c @ 247])
+ #4  0xf7638f16 ((0006bf16) win32k.sys!NtUserLoadKeyboardLayoutEx+1a6 [h:\project\reactos\win32ss\user\ntuser\kbdlayout.c @ 650])
+ #5  0x8054d99b ((0014d99b) ntoskrnl.exe!KiSystemCallTrampoline+1b [h:\project\reactos\ntoskrnl\include\internal\i386\ke.h @ 766])
+ #6  0x8054b638 ((0014b638) ntoskrnl.exe!KiSystemServiceHandler+278 [h:\project\reactos\ntoskrnl\ke\i386\traphdlr.c @ 1836])
+ #7  0x80403da3 ((00003da3) ntoskrnl.exe!KiFastCallEntry+8c)
+ ```
+
+ #0 and #7 don't have source code information because they were written in assembly.
```

Comment:
```suggestion
 #0 and #7 don't have source code information because they are written in assembly.
```
With `were` you imply that the aforementioned functions were written in Assembly in the past and they're not anymore.